### PR TITLE
Grammar cleanup

### PR DESCRIPTION
**Pull request by jstolarek**
Addresses #426.

Below is a parser refactoring that I have been working on for the past week or so. I mostly tried to compactify the rules by relying on Menhir combinators. So for example instead of productions like:

```
foos:
| foo      { $1       }
| foo foos { $1 :: $2 }
```

we now have:

```
foos:
| foo+      { $1 }
```

As a result `parser.mly` is now about 100 lines shorter and about 13% smaller in terms of file size in bytes. This is as far as I was able to go without altering the structure of Links AST and touching further stages of compilation pipeline. More improvements are possible if we did so, but I personally don't yet feel competent enough to do this. A good example of a possible improvement are the three production rules for `INSERT` already discussed in https://github.com/links-lang/links/issues/426#issuecomment-442583867:

```
| INSERT exp VALUES LBRACKET LPAREN labeled_exps RPAREN RBRACKET
  { with_pos $loc (`DBInsert ($2, labels $6,
     with_pos $loc($6) (`ListLit ([with_pos $loc($6) (`RecordLit ($6, None))], None)),
    None)) }

| INSERT exp VALUES LBRACKET LPAREN RPAREN RBRACKET RETURNING VARIABLE
  { with_pos $loc (`DBInsert ($2, [],
     with_pos $loc (`ListLit ([with_pos $loc (`RecordLit ([], None))], None)),
    Some (with_pos $loc($9) (`Constant (`String $9))))) }

| INSERT exp VALUES LBRACKET LPAREN labeled_exps RPAREN RBRACKET RETURNING VARIABLE
  { with_pos $loc (`DBInsert ($2, labels $6,
     with_pos $loc($6) (`ListLit ([with_pos $loc($6) (`RecordLit ($6, None))], None)),
    Some (with_pos $loc (`Constant (`String $10))))) }
```

These can be collapsed into a single rule:

```
| INSERT exp VALUES LBRACKET LPAREN loption(labeled_exps) RPAREN RBRACKET
         preceded(RETURNING, VARIABLE)?
  { with_pos $loc (`DBInsert ($2, labels $6,
     with_pos $loc($6) (`ListLit ([with_pos $loc($6) (`RecordLit ($6, None))], None)),
    OptionUtils.opt_map (fun v -> with_pos $loc (`Constant (`String v))) $9)) }
```
But this requries that later stages of compilation pipeline catch a case where `labeled_exps` is an empty list and there is no `RETURNING VARIABLE`.

There is one more thing I am tempted to do, but I first need to know what others think. Currently `parser.mly` is formatted in such a way that productions are placed on the left and semantic actions are placed in the same line on the right, starting in column 63. This leads to extremely long lines (as much as 140 columns). I am often working with two source files opened side by side, but with `parser.mly` formatted the way it is now this is impossible. Similalrly, I suspect many of you will experience problems viewing changes in this PR, because lines are too long to be displayed side by side and they need to be wrapped, which makes viewing difficult. I would like to reformat `parser.mly` so that semantic actions are placed below productions whenever necessary. An example of this is above - the three INSERT rules are reformatted using this style for the purpose of this comment. In the source code they currently look like this:

```
| INSERT exp VALUES
  LBRACKET LPAREN labeled_exps RPAREN RBRACKET                 { with_pos $loc (`DBInsert ($2,
                                                                          labels $6,
                                                                          with_pos $loc($6)
                                                                                   (`ListLit ([with_pos $loc($6)
                                                                                                        (`RecordLit ($6, None))], None)),
                                                                          None)) }
| INSERT exp VALUES
  LBRACKET LPAREN RPAREN RBRACKET
  RETURNING VARIABLE                                           { with_pos $loc (`DBInsert ($2,
                                                                          [],
                                                                          (with_pos $loc (`ListLit ([with_pos $loc (`RecordLit ([], None))],
                                                                                                    None))),
                                                                          Some (with_pos $loc($9) (`Constant (`String $9))))) }
| INSERT exp VALUES
  LBRACKET LPAREN labeled_exps RPAREN RBRACKET
  RETURNING VARIABLE                                           { with_pos $loc (`DBInsert ($2,
                                                                          labels $6,
                                                                          (with_pos $loc($6) (`ListLit ([with_pos $loc($6)
                                                                                                                  (`RecordLit ($6, None))],
                                                                                                        None))),
                                                                          Some (with_pos $loc (`Constant (`String $10))))) }

```
which, I think, is horrible to read.  Are there any strong opinions for or against such reformatting?

Finally, there is this note in `parser.mly`:

```
/* FIXME:
 *
 * recursive row vars shouldn't be restricted to vfields.
 */
row_var:
| nonrec_row_var                                               { $1 }
| LPAREN MU VARIABLE DOT vfields RPAREN                        { `Recursive ($3, $5) }
```

What should the other allowed productions be? We can now parameterize production rules so this should be easy to fix.

In general, it would be nice if these kind of TODO/FIXME notes were avoided and appropriate tickets with informative description be created instead. (I am guessing many such notes come from times when there was no bugtracker?) I believe this would greatly increase the chances that such TODOs ever get addressed.

### CODE REVIEW

**Comment by jstolarek** [comment]
I will merge this on Monday if there are no comments/complaints.

File: core/parser.mly [hunk 1]
```diff
@@ -1013,14 +968,11 @@ binding:
...
                                                                                   None)) }
-| typedecl SEMICOLON                                           { $1 }
 | typed_handler_binding                                        { with_pos $loc (`Handler $1) }
-| links_module                                                 { $1 }
-| alien_block                                                  { $1 }
-| links_open                                                   { $1 }
+| typedecl SEMICOLON | links_module | alien_block | links_open { $1 }
 
 bindings:
-| binding                                                      { [$1] }
```
Comment (frank-emrich) [inline core/parser.mly]: shouldn't this collapse to a single  rule  ` bindings -> binding+`  ?

File: core/parser.mly [hunk 2]
```diff
@@ -1013,14 +968,11 @@ binding:
...
+| binding                                                      { [$1]      }
 | bindings binding                                             { $1 @ [$2] }
```
Comment (frank-emrich) [inline core/parser.mly]: shouldn't this collapse to a single  rule  ` bindings -> binding+`  ?

(sorry, the original comment was on the wrong line)

Comment (jstolarek) [reply core/parser.mly]: I will reply in main discussion - this is too long for an inline comment.

**Comment by jstolarek** [comment]
Replying to Frank's inline comment:

> shouldn't this collapse to a single rule bindings -> binding+

I wish this was the case, but no - doing so leads to a shift/reduce conflict. I believe this has to do with the current definition being left-recursive, whereas `binding+` definition would be right-recursive. Here is what Menhir says about the conflict:

```
LBRACE block_contents RBRACE 
       (?)

** In state 922, looking ahead at VARIABLE, shifting is permitted
** because of the following sub-derivation:

bindings exp SEMICOLON 
nonempty_list(binding) 
binding nonempty_list(binding) 
        binding 
        exp SEMICOLON 
        lens_expression 
        database_expression 
        table_expression 
        formlet_expression 
        escape_expression 
        iteration_expression 
        case_expression 
        conditional_expression 
        session_expression 
        db_expression 
        typed_expression 
        logical_expression 
        infixl_0 
        infixr_0 
        infixl_1 
        infixr_1 
        infixl_2 
        infixr_2 
        infixl_3 
        infixr_3 
        infixl_4 
        infixr_4 
        infixr_5 EQUALSTILDE regex 
        infixl_6 
        infixr_6 
        infixl_7 
        infixr_7 
        infixl_8 
        infixr_8 
        infixl_9 
        infixr_9 
        unary_expression 
        postfix_expression 
        primary_expression 
        atomic_expression 
        . VARIABLE 

** In state 922, looking ahead at VARIABLE, reducing production
** nonempty_list(binding) -> binding 
** is permitted because of the following sub-derivation:

bindings exp SEMICOLON // lookahead token appears because exp can begin with VARIABLE
nonempty_list(binding) // lookahead token is inherited
binding .
```

So for a right-recursive definition, after reducing one `binding` production and encountering `VARIABLE` token it is not clear whether that token is another binding (then shift in `bindings` production) or an `exp` (then reduce `bindings`). I am not 100% sure why this problem does not arise for a left-recursive definition. I presume that for such a definition all such ambiguities are resolved in favour of reducing as many bindings to the right as possible.

There are several more such productions, but not many. Perhaps there is a way to restructure the grammar in a way that avoids this kind of ambiguities, but I admit I don't know Links well enough to do it yet.

**Comment by jstolarek** [comment]
Merged as 951270883de0fb90988845c03ff91f13224a6a8a.

I will submit a separate PR with a proposed refactoring of `parser.mly` to 80-column format.

**Pull request closed without merging**